### PR TITLE
[15.0][IMP] subscription_oca: Tests, refactoring

### PR DESCRIPTION
**Pull request by ilyasProgrammer**

### CODE REVIEW

**Comment by francesco-ooops** [comment]
@carlos-domatix what do you think about tests?

@simahawk implementation of https://github.com/OCA/contract/pull/971#pullrequestreview-1639482033

File: subscription_oca/tests/test_subscription_oca.py [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+
+from odoo import fields
+from odoo.tests import TransactionCase
+
+
+class TestSubscriptionOCA(TransactionCase):
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        cls.pricelist = cls.env["product.pricelist"].create(
```
Comment (simahawk) [inline subscription_oca/tests/test_subscription_oca.py:15]:
```suggestion
        cls.env = cls.env(context=dict(cls.env.context, tracking_disable=True))
        cls.pricelist = cls.env["product.pricelist"].create(
```

File: subscription_oca/tests/test_subscription_oca.py [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+                "name": "10% Tax incl",
+                "amount_type": "percent",
+                "amount": 10,
+                "price_include": True,
+            }
+        )
+        cls.product_1 = cls.env.ref("product.product_product_1")
+        cls.product_1.taxes_id = [(6, 0, cls.tax_10pc_incl.ids)]
+        cls.product_2 = cls.env.ref("product.product_product_2")
+        cls.country_id = cls.env["res.country"].search([], limit=1)
```
Comment (simahawk) [inline subscription_oca/tests/test_subscription_oca.py]:
```suggestion
        cls.country = cls.env["res.country"].search([], limit=1)
```
to do for all vars that are records and not IDs

File: subscription_oca/tests/test_subscription_oca.py [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+                "product_id": cls.product_1.id,
+            }
+        )
+        cls.close_reason_id = cls.env["sale.subscription.close.reason"].create(
+            {
+                "name": "Test Close Reason",
+            }
+        )
+
+    def test_all_subscription_oca(self):
```
Comment (simahawk) [inline subscription_oca/tests/test_subscription_oca.py]: Any major reason to have just one biiiig test?
Still better than nothing of course :)
If you don't have time, add a TODO for splitting it

File: subscription_oca/tests/test_subscription_oca.py [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+                            "product_id": self.product_1.id,
+                            "product_uom_qty": 2,
+                            "product_uom": self.product_1.uom_id.id,
+                            "price_unit": self.product_1.list_price,
+                        },
+                    )
+                ],
+            }
+        )
+        so._compute_subscriptions_count()
```
Comment (simahawk) [inline subscription_oca/tests/test_subscription_oca.py:117]: compute methods should be computed automatically. IF you need need this maybe there's a problem

File: subscription_oca/tests/test_subscription_oca.py [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+        self.assertEqual(self.sub_line_id.discount, 0)
+        res = self.sub_line_id._get_display_price(self.product_2)
+        self.assertEqual(res, 38.25)
+        sol_res = self.sub_line_id._prepare_sale_order_line()
+        self.assertIsInstance(sol_res, dict)
+        move_res = self.sub_line_id._prepare_account_move_line()
+        self.assertIsInstance(move_res, dict)
+
+        # sale.subscription
+        self.sub_id.cron_subscription_management()
```
Comment (simahawk) [inline subscription_oca/tests/test_subscription_oca.py]: this should go to one or more cron specific tests

File: subscription_oca/tests/test_subscription_oca.py [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+        )
+        self.sub_id.onchange_partner_id_fpos()
+        self.assertFalse(self.sub_id.fiscal_position_id)
+        res = self.sub_id.action_close_subscription()
+        self.assertEqual(res["type"], "ir.actions.act_window")
+
+        # sale.subscription.stage
+        self.stage_id._check_lot_product()  # should not raise
+
+        # sale.subscription.template
```
Comment (simahawk) [inline subscription_oca/tests/test_subscription_oca.py]: testing a template? :)

**Review by simahawk — CHANGES_REQUESTED** [review]
Thanks a lot for the effort :)

**Comment by ilyasProgrammer** [comment]
@simahawk Thanks for the review. Please have a look again.

**Comment by carlos-domatix** [comment]
Thanks a lot for the effort. I see it well but it has 70% coverage https://app.codecov.io/gh/OCA/contract/tree/15.0-subscription_oca-tests/, I don't know if it is enough for the repo, but at least it is a good base.

**Review by simahawk — APPROVED** [review]

**Pull request merged**